Add feature flag for enumerated ARIA attributes and new ARIA IDL

rahimabdi opened the pull request:
#### 6677c35087826e7c1e75599d5613bc6b181e5bf6
<pre>
Add feature flag for enumerated ARIA attributes and new ARIA IDL
<a href="https://bugs.webkit.org/show_bug.cgi?id=296851">https://bugs.webkit.org/show_bug.cgi?id=296851</a>
<a href="https://rdar.apple.com/157952768">rdar://157952768</a>

Reviewed by Tyler Wilcock and Ryosuke Niwa.

Adds &apos;status: developer&apos; feature flag for enumerated ARIA attributes: <a href="https://github.com/w3c/aria/pull/2484.">https://github.com/w3c/aria/pull/2484.</a>

The ARIA spec intends to convert the following attributes to be enumerated, i.e., content attributes
with permissible keywords and invalid/missing value defaults:
- aria-atomic
- aria-autocomplete
- aria-busy
- aria-checked
- aria-current
- aria-disabled
- aria-expanded
- aria-haspopup
- aria-hidden
- aria-invalid
- aria-live
- aria-modal
- aria-multiline
- aria-multiselectable
- aria-orientation
- aria-pressed
- aria-readonly
- aria-required
- aria-selected
- aria-sort

This PR changes ARIA&apos;s IDL to use [ReflectSetter] for enumerated attributes in order to implement the desired reflection
behavior behind a flag using manual getters; this also allows maintaining current ARIA IDL behavior when the flag is disabled.
When ARIA PR #2484 lands, ARIA&apos;s IDL should be updated to use [Reflect, Enumerated] for automatic JS bindings codegen
that doesn&apos;t need manual code.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/accessibility/AriaAttributes.idl:
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::ariaAtomic const):
(WebCore::Element::ariaAutoComplete const):
(WebCore::Element::ariaBusy const):
(WebCore::Element::ariaChecked const):
(WebCore::Element::ariaCurrent const):
(WebCore::Element::ariaDisabled const):
(WebCore::Element::ariaExpanded const):
(WebCore::Element::ariaHasPopup const):
(WebCore::Element::ariaHidden const):
(WebCore::Element::ariaInvalid const):
(WebCore::Element::ariaLive const):
(WebCore::Element::ariaModal const):
(WebCore::Element::ariaMultiLine const):
(WebCore::Element::ariaMultiSelectable const):
(WebCore::Element::ariaOrientation const):
(WebCore::Element::ariaPressed const):
(WebCore::Element::ariaReadOnly const):
(WebCore::Element::ariaRequired const):
(WebCore::Element::ariaSelected const):
(WebCore::Element::ariaSort const):
* Tools/DumpRenderTree/TestOptions.cpp:
* Tools/WebKitTestRunner/TestOptions.cpp:

Canonical link: <a href="https://commits.webkit.org/299159@main">https://commits.webkit.org/299159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69ae556aff5357e0d8f76e6055fd7f6dfe900096

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118043 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37721 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28357 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124193 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70079 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119921 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38414 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46303 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89581 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59182 "Too many flaky failures: css3/filters/drop-shadow-current-color.html, fast/canvas/offscreen-no-script-context-crash.html, http/tests/permissions/storage-access-permissions-query.html, http/tests/resourceLoadStatistics/only-accept-first-party-cookies-with-third-party-cookie-blocking.html, imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative.html, imported/w3c/web-platform-tests/trusted-types/set-event-handlers-content-attributes.tentative.html, imported/w3c/web-platform-tests/wasm/core/simd/simd_lane.wast.js.html, js/regexp-many-brackets.html, storage/indexeddb/getdatabases.html, storage/indexeddb/intversion-omit-parameter-private.html, storage/indexeddb/modern/opendatabase-success-after-versionchange-private.html, webgl/2.0.0/conformance2/textures/image/tex-3d-rgba8ui-rgba_integer-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_blob/tex-2d-srgb8-rgb-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_blob/tex-3d-rgb565-rgb-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_blob/tex-3d-rgb565-rgb-unsigned_short_5_6_5.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_blob/tex-3d-rgba4-rgba-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_image/tex-3d-rgb5_a1-rgba-unsigned_short_5_5_5_1.html, webgl/2.0.0/conformance2/textures/image_data/tex-3d-rgba16f-rgba-half_float.html, webgl/2.0.0/conformance2/textures/webgl_canvas/tex-2d-rgb5_a1-rgba-unsigned_byte.html, webgl/2.0.0/conformance2/textures/webgl_canvas/tex-2d-rgb5_a1-rgba-unsigned_short_5_5_5_1.html, webgl/2.0.y/conformance/extensions/ext-color-buffer-half-float.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120995 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30598 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105848 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70073 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29666 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23965 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67858 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110154 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100021 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24143 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127274 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116553 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44946 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33869 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98256 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45307 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102072 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98042 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24939 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43437 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21427 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41382 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44818 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50492 "Built successfully") | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145251 "Hash 69ae556a for PR 48937 does not build (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44278 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/145251 "Hash 69ae556a for PR 48937 does not build (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47623 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45967 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/dom/Element.h:
```diff
@@ -222,6 +222,31 @@ class Element : public ContainerNode {
+    enum class AriaReadOnlyValues { True, False };
+    enum class AriaRequiredValues { True, False };
+    enum class AriaSelectedValues { True, False };
+    enum class AriaSortValues { Ascending, Descending, Other, None };
```
We should give types to these, as otherwise they'll default to the too-large value of `int`. Let's use `bool` for the two-value enums, and `uint8_t` for the three-value enums.

`enum class AriaAtomicValues : bool { True, False };`
`enum class AriaAutoCompleteValues : uint8_t { Inline, List, Both, None };`

rahimabdi replied:
Got it, thanks.

lukewarlow commented on Source/WebCore/dom/Element.h:
```diff
@@ -222,6 +222,31 @@ class Element : public ContainerNode {
     HTMLElement* topmostPopoverAncestor(TopLayerElementType topLayerType);
 
+#if ENABLE(ARIA_IDL_REFLECTION)
+    // aria-current uses a custom ariaCurrent() getter via [ReflectSetter] rather than [Reflect, Enumerated] due to complex reflection.
```
Maybe reword this, the getter isn't "via" ReflectSetter it just uses that + a defined getter.

rahimabdi replied:
Agreed, thanks!

rahimabdi replied:
New comment: 

> // aria-current uses [ReflectSetter] with a custom getter rather than [Reflect, Enumerated] due to complex reflection.

annevk commented on Source/WebCore/accessibility/AriaAttributes.idl:
```diff
@@ -25,6 +25,59 @@
+    [CEReactions=Needed, Reflect="aria-valuemax"]         attribute DOMString? ariaValueMax;
+    [CEReactions=Needed, Reflect="aria-valuemin"]         attribute DOMString? ariaValueMin;
+    [CEReactions=Needed, Reflect="aria-valuenow"]         attribute DOMString? ariaValueNow;
+    [CEReactions=Needed, Reflect="aria-valuetext"]        attribute DOMString? ariaValueText;
```
This one is the same as the one in the else block. If we're going with this approach while we transition (I guess this is hard to do with a pure runtime setting?) I don't think we should duplicate identical entries.

rahimabdi replied:
Agreed.

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -6352,4 +6352,34 @@ void Element::setNumericAttribute(const QualifiedName& attributeName, double val
+        return timeValue.get();
+    if (equalLettersIgnoringASCIICase(value, "true"_s))
+        return trueValue.get();
+    return falseValue.get();
```
Can you explain why this can't be done with an enum?

rahimabdi replied:
Both `aria-invalid` and `aria-current` support the empty string as a keyword which maps to "false" (along with the keyword "false"). Since the empty string is a synonym of "false", it cannot be listed in the enum as a unique value, and it also isn't captured by the invalid value default (which is "true") so it needs a manual getter.

Once WebKitIDL supports an "empty value default", these attributes can be updated to use [Reflect, Enumerated] like the others, i.e.,:

[MissingValueDefault="false", InvalidValueDefault="true", EmptyValueDefault="false"] enum ElementAriaInvalidValues { "grammar", "spelling", "true", "false" };

CC @lukewarlow

annevk commented on Source/WebCore/accessibility/AriaAttributes.idl:
```diff
@@ -76,4 +129,27 @@ interface mixin AriaAttributes {
+[MissingValueDefault="false", InvalidValueDefault="false"] enum ElementAriaRequiredValues { "true", "false" };
+enum ElementAriaSelectedValues { "true", "false" };
+[MissingValueDefault="false", InvalidValueDefault="false"] enum ElementAriaSortValues { "ascending", "descending", "other", "none" };
+#endif // ENABLE_ARIA_IDL_REFLECTION
```
Let's include newlines at the end of a file.

rniwa commented on Source/WebCore/accessibility/AriaAttributes.idl:
```diff
@@ -25,6 +25,59 @@
 interface mixin AriaAttributes {
+#if defined(ENABLE_ARIA_IDL_REFLECTION) && ENABLE_ARIA_IDL_REFLECTION
+    [CEReactions=Needed, Reflect="aria-activedescendant"] attribute Element? ariaActiveDescendantElement;
+    [CEReactions=Needed, Reflect="aria-atomic", Enumerated=ElementAriaAtomicValues]                 attribute DOMString? ariaAtomic;
```
Please consistently use single space as a delimiter between [~] and attribute throughout the file.

rahimabdi replied:
Fixed.

rniwa commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -561,6 +561,21 @@ ApplePayEnabled:
   sharedPreferenceForWebProcess: true
   richJavaScript: true
 
+AriaIDLNewReflection:
```
ARIA should be capitalized. I don't think "IDL" is necessary but we should be explicitly about the fac this isa bout enumerated attribute so `EnumeratedARIAAttributeReflectionEnabled` would be a better name.

rahimabdi replied:
Agreed, changed to `EnumeratedARIAAttributeReflectionEnabled`. Thank you!

rniwa commented on Source/WebCore/accessibility/AriaAttributes.idl:
```diff
@@ -25,6 +25,59 @@
 // https://w3c.github.io/aria/#idl-interface
 interface mixin AriaAttributes {
+#if defined(ENABLE_ARIA_IDL_REFLECTION) && ENABLE_ARIA_IDL_REFLECTION
+    [CEReactions=Needed, Reflect="aria-activedescendant"] attribute Element? ariaActiveDescendantElement;
```
This is identical to when this feature is disabled. Why do we need to re-implement it here?

rahimabdi replied:
@annevk left a similar comment, I've updated to specify only changed attributes in the flag.

rahimabdi committed:
Add feature flag for enumerated ARIA attributes and new ARIA IDL
https://bugs.webkit.org/show_bug.cgi?id=296851
rdar://157952768

Reviewed by Tyler Wilcock and Ryosuke Niwa.

Adds 'status: developer' feature flag for enumerated ARIA attributes: https://github.com/w3c/aria/pull/2484.

The ARIA spec intends to convert the following attributes to be enumerated, i.e., content attributes
with permissible keywords and invalid/missing value defaults:
- aria-atomic
- aria-autocomplete
- aria-busy
- aria-checked
- aria-current
- aria-disabled
- aria-expanded
- aria-haspopup
- aria-hidden
- aria-invalid
- aria-live
- aria-modal
- aria-multiline
- aria-multiselectable
- aria-orientation
- aria-pressed
- aria-readonly
- aria-required
- aria-selected
- aria-sort

This PR changes ARIA's IDL to use [ReflectSetter] for enumerated attributes in order to implement the desired reflection
behavior behind a flag using manual getters; this also allows maintaining current ARIA IDL behavior when the flag is disabled.
When ARIA PR #2484 lands, ARIA's IDL should be updated to use [Reflect, Enumerated] for automatic JS bindings codegen
that doesn't need manual code.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/accessibility/AriaAttributes.idl:
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::ariaAtomic const):
(WebCore::Element::ariaAutoComplete const):
(WebCore::Element::ariaBusy const):
(WebCore::Element::ariaChecked const):
(WebCore::Element::ariaCurrent const):
(WebCore::Element::ariaDisabled const):
(WebCore::Element::ariaExpanded const):
(WebCore::Element::ariaHasPopup const):
(WebCore::Element::ariaHidden const):
(WebCore::Element::ariaInvalid const):
(WebCore::Element::ariaLive const):
(WebCore::Element::ariaModal const):
(WebCore::Element::ariaMultiLine const):
(WebCore::Element::ariaMultiSelectable const):
(WebCore::Element::ariaOrientation const):
(WebCore::Element::ariaPressed const):
(WebCore::Element::ariaReadOnly const):
(WebCore::Element::ariaRequired const):
(WebCore::Element::ariaSelected const):
(WebCore::Element::ariaSort const):
* Tools/DumpRenderTree/TestOptions.cpp:
* Tools/WebKitTestRunner/TestOptions.cpp:

Canonical link: https://commits.webkit.org/299159@main
